sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,752 @@
+
+  // NOTE: We recommend setting clean session to false in order to receive any pending messages
+  mqtt_connect_options.cleansession = false;
+  mqtt_connect_options.keepAliveInterval = AZ_IOT_DEFAULT_MQTT_CONNECT_KEEPALIVE_SECONDS;
+
+  // Get the MQTT user name used to connect to IoT Hub
+  if (az_failed(
+          rc = az_iot_hub_client_get_user_name_with_model_id(
+              &client, model_id, mqtt_username, sizeof(mqtt_username), NULL)))
+
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  printf("%s\r\n", mqtt_username);
+
+  // This sample uses X509 authentication so the password field is set to NULL
+  mqtt_connect_options.username = mqtt_username;
+  mqtt_connect_options.password = NULL;
+
+  // Set the device cert for tls mutual authentication
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  if (*x509_trust_pem_file != '\0')
+  {
+    mqtt_ssl_options.trustStore = (char*)x509_trust_pem_file;
+  }
+
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  // Connect to IoT Hub
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int subscribe()
+{
+  int rc;
+
+  // Subscribe to the methods topic. Messages received on this topic are methods to be invoked
+  // on the device.
+  if ((rc = MQTTClient_subscribe(mqtt_client, AZ_IOT_HUB_CLIENT_METHODS_SUBSCRIBE_TOPIC, 1))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe to twin patch topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  // Subscribe to the desired properties PATCH topic. Messages received on this topic will be
+  // updates to the desired properties.
+  if ((rc = MQTTClient_subscribe(mqtt_client, AZ_IOT_HUB_CLIENT_TWIN_PATCH_SUBSCRIBE_TOPIC, 1))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe to twin patch topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  // Subscribe to the twin response topic. Messages received on this topic will be response statuses
+  // from published reported properties or the requested twin document from twin GET publish
+  // messages
+  if ((rc = MQTTClient_subscribe(mqtt_client, AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_SUBSCRIBE_TOPIC, 1))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe to twin response topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+// Send JSON formatted telemetry messages
+static int send_telemetry_messages()
+{
+  int rc;
+
+  if (az_failed(
+          rc = az_iot_hub_client_telemetry_get_publish_topic(
+              &client, NULL, telemetry_topic, sizeof(telemetry_topic), NULL)))
+  {
+    return rc;
+  }
+
+  az_json_builder json_builder;
+  az_result result;
+  result = az_json_builder_init(&json_builder, AZ_SPAN_FROM_BUFFER(telemetry_payload), NULL);
```

Comment:
There's a bug here because you're building up the payload once, but if there was a `current_device_temp` changed because of a property change in between one of these sends then you would not be sending the correct value back.

Ideally this `send_telemetry_messages()` would loop infinitely (with a longer sleep than 1 second to avoid taking too much quota on a hub) so it'd reflect the value actually changing in response to properties incoming.  But the problem is your `getchar()` in your main() is blocking, so without spinnang a thread in this sample (don't do that!) there's not a good way to let user ext program celanly and have infinite loop of send telemetry forever.  

You could alternately take the approach of SDK v1 samples of <printf("Control-C to break")> (and leave the cleanup code after this part in in mostly for copy/pasters).

?

Reply:
ah crap you're right thanks. I had meant to update that.

So this sample uses the threaded version of Paho so the `on_received` is the callback invoked for any incoming messages. We have a backlog item to change them to the non-threaded but that will take a bit of work.

I can change it to increase the interval and build up the payload each time. That shouldn't be an issue.

Reply:
Since you're using the same variable on 2 threads you'll need to ensure the variable is not optimized by the compiler (`volatile`) then use some of the interlocked APIs (or OS-level locks) to change the values.

Reply:
It has been converted to synchronous to try and alleviate many of these problems